### ISSUE TITLE
Add Magnetic stripe card reader for card payments

### ISSUE THREAD

**Issue by mwillis72** [opensourcepos/opensourcepos#2892]
### Background information

__IMPORTANT: If you choose to ignore this issue report template, your issue will be closed as we cannot help without the requested information.__

Please make sure you tick (add an x between the square brackets with no spaces) the following check boxes:

- [] Reporting an issue of an unmodified OSPOS installation
- [] Checked [open and closed issues](https://github.com/opensourcepos/opensourcepos/issues?utf8=%E2%9C%93&q=is%3Aissue) and no similar issue was already reported (please make sure you searched!)
- [] Read [README](https://github.com/opensourcepos/opensourcepos/blob/master/README.md), [WHATS_NEW](https://github.com/opensourcepos/opensourcepos/blob/master/WHATS_NEW.txt), [INSTALL.md](https://github.com/opensourcepos/opensourcepos/blob/master/INSTALL.md) and [UPGRADE](https://github.com/opensourcepos/opensourcepos/blob/master/UPGRADE.txt)
- [] Read the [FAQ](https://github.com/opensourcepos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP extensions installed)
- [] Read the [wiki](https://github.com/opensourcepos/opensourcepos/wiki)
- [] Executed any database upgrade scripts if an upgrade pre 3.0.0 (e.g. database/2.4_to_3.0.sql)
- [] Aware the installation code is in [bintray](https://bintray.com/jekkos/opensourcepos/opensourcepos/view/files?sort=updated&order=asc#files) (see README), and [GitHub master](https://github.com/opensourcepos/opensourcepos/tree/master) is for [developers only](https://github.com/opensourcepos/opensourcepos/wiki/Development-setup) and therefore not complete nor stable

### Installation information

- OSPOS version is: 
- OSPOS git commit hash is: 
- PHP version is: 
- MySQL or MariaDB version is: 
- OS and version is: 
- WebServer is: 
- Selected language is: 
- (If applicable) Docker installation: 
- (If applicable) Installation package for the LAMP/LEMP stack is: 

### Issue / Bug  / Question / New Feature

Please write your issue here. If a bug, please make sure to provide as much information as possible including configuration settings (e.g. Decimals set, Tax mode), language and steps to reproduce the bug.

**Comment by jekkos** [comment]
@mwillis72 your request is very vague. Please elaborate a bit on what card reader, how you want to connect it to the browser and how the payments should be processed.

**Comment by mwillis72** [comment]
I mean Credit card and debit card. If a customer wants to pay with credit
card or debit card, it is striped through the magnetic stripe reader(which
is connected to the OSPOS system/computer), the card info is detected by
and displayed in ospos at checkout (when adding payment type as debit or
credit) and it is then charged. After the card is charged, then a
receipt/invoice can be printed for the customer! Hope its clearer

**Comment by mwillis72** [comment]
Insertion Magnetic card reader and flutterwave payment gateway if possible

**Comment by mckaygerhard** [comment]
firts of all the issue feature request are not filled well..  only the title was provided.. rest are template only

@mwillis72  that stripe always have a SDK to integrate to.. this software runs as client-server web application.. so the possible ways are webservices or jssockets.. if the magnetic card reader does not provide SDK integration and documentations never will be possible.. (taking in consideration all the software are made for the crap of windo, with closed source things.. for sure 60% of vendors do not provide good SDK or API's)

is the same case of the fiscal printers.. only bluetooh printer are good to use!

**Comment by mwillis72** [comment]
@mckaygerhard Is there any stripe reader with sdk  that you can recommend for me to integrate in this software???? Because I seriously need the feature to be integrated

**Comment by professormunki** [comment]
sumUp provide an api and sdk for their hardware and internet payment gateway, they're quite a popular vendor in Ireland. 
Adyen also make the relevant information available through their website.

These are emv readers, if you're not trying to use this type of gateway but rather a magnetic stripe 'store card' try the hardware manufacturer.

These are the only two gateways I've needed to look at so far, the methods employed by both providers are similar (using their hardware to handle verification, encryption and payment) receiving a transaction/register id and value from the POS and returning success or error code. 

The differences between the devices offered and the provider requirements would make a global module quite complicated ,unless limiting the capabilities to the most common use cases, eg. payment and refund. (although comparing end of day (z reads) with the cash-up totals would be nice, most of the ospos user error I've seen has been entering payments by card as cash payments or vice-versa). For SumUp devices there are two ways of interacting with the device from the POS (depending on the device) : web API (also an option with Adyen) or bluetooth with their sdk (android and ios)

I suspect there are a myriad of apis and sdks available, probably too many. All absolutely strict about their requirements. 
If you're already using a payment gateway for credit/debit cards at point of sale, check with them first. 

I'm certainly no expert on PHP but I expect a hook/helper attached to the payment submission button triggering the card terminal transaction and waiting for a response would be the most rational place to inject it. The function would theoretically be Passing the transaction id and value to the device (by preconfigured pairing) along with the command for payment or refund (depending on the value of the transaction) and waiting for: timeout (error: $device not connected) || failed (error:rejected, not authorized, insufficient funds etc) || success (submit form, update $(this))

In short, it should be relatively easy to do on a case by case basis provided that the APIs or SDKs are available but catering for all options would be a config nightmare.
Just my two cents.

**Comment by jekkos** [comment]
Well yes it would fit perfectly on the third party plugin infrastructure @objecttothis started to work on. Basically when a sale complete you would send an event containing relevant sale data, and then have the payment processor attached through that interface to post process the sale.

**Comment by objecttothis** [comment]
@jekkos agreed that this would integrate well with the integrations enhancement.  We just need to get that sorted out and merged with the master.  The current implementation works (I'm using it in production), but it's basic and @jekkos and I think we may have found a better alternative event generator.  I will just need some help with the design and scaffolding of it.

**Comment by stellarcheese** [comment]
Was this ever done?

**Comment by objecttothis** [comment]
> 
> 
> Was this ever done?

The third party integrations has not yet been integrated into OSPOS, so no.  Currently I only have two days per week that I can work on OSPOS stuff and most of that in the last few months has been taken up with an online payment method for my website.  That's ending soon and I can begin focusing more on finishing up code review changes on the CSV Imports of items.  Once that is finished and merged into the master I can turn all of my attention on getting the third-party integration framework implemented.  I'll need to work with @jekkos and maybe @daN4cat on a design that everyone is happy with, since my current implementation is rudimentary (but does work).  Once third-party integrations is implemented in the master, you or someone will be free to develop an integration allowing communication with magnetic stripe card readers.

Of course this is all assuming that magnetic stripe card readers should be developed as a third-party integration rather than a core functionality (like a secondary screen, receipt printer, etc).  That's another issue that should be decided.

**Comment by stellarcheese** [comment]
Are you willing to send me the code you are using to integrate?  Rudimentary is fine but I need to speed up the check out process as I am seeing more business every week. 

Willing to take as is and help the process move forward.

I myself run my repair shop by myself but am hiring soon so will have some time to help where I can.

**Comment by objecttothis** [comment]
@stellarcheese The code is in Pull Request #2619 So you are welcome to integrate that and build the third-party integration into that, but keep in mind that it's very likely that the third-party integrations framework will likely change, so you may need to adapt the code for the stripe reader integration to fit the final product.

**Comment by doremonx** [comment]
9677
